src/write_prometheus.c

```diff
@@ -0,0 +1,829 @@
+
+  int status = MHD_queue_response(connection, MHD_HTTP_OK, res);
+
+  MHD_destroy_response(res);
+  PROTOBUF_C_BUFFER_SIMPLE_CLEAR(&simple);
+  return status;
+}
+
+/*
+ * Functions for manipulating the global state in "metrics". This is organized
+ * in two tiers: the global "metrics" tree holds "metric families", which are
+ * identified by a name (a string). Each metric family has one or more
+ * "metrics", which are identified by a unique set of key-value-pairs. For
+ * example:
+ *
+ * collectd_cpu_total
+ *   {cpu="0",type="idle"}
+ *   {cpu="0",type="user"}
+ *   ...
+ * collectd_memory
+ *   {memory="used"}
+ *   {memory="free"}
+ *   ...
+ * {{{ */
+/* label_pair_destroy frees the memory used by a label pair. */
+static void label_pair_destroy(Io__Prometheus__Client__LabelPair *msg) {
+  if (msg == NULL)
+    return;
+
+  sfree(msg->name);
+  sfree(msg->value);
+
+  sfree(msg);
+}
+
+/* label_pair_clone allocates and initializes a new label pair. */
+static Io__Prometheus__Client__LabelPair *
+label_pair_clone(Io__Prometheus__Client__LabelPair const *orig) {
+  Io__Prometheus__Client__LabelPair *copy = calloc(1, sizeof(*copy));
+  if (copy == NULL)
+    return NULL;
+  io__prometheus__client__label_pair__init(copy);
+
+  copy->name = strdup(orig->name);
+  copy->value = strdup(orig->value);
+  if ((copy->name == NULL) || (copy->value == NULL)) {
+    label_pair_destroy(copy);
+    return NULL;
+  }
+
+  return copy;
+}
+
+/* metric_destroy frees the memory used by a metric. */
+static void metric_destroy(Io__Prometheus__Client__Metric *msg) {
+  if (msg == NULL)
+    return;
+
+  for (size_t i = 0; i < msg->n_label; i++) {
+    label_pair_destroy(msg->label[i]);
+  }
+  sfree(msg->label);
+
+  sfree(msg->gauge);
+  sfree(msg->counter);
+
+  sfree(msg);
+}
+
+/* metric_cmp compares two metrics. It's prototype makes it easy to use with
+ * qsort(3) and bsearch(3). */
+static int metric_cmp(void const *a, void const *b) {
+  Io__Prometheus__Client__Metric const *m_a =
+      *((Io__Prometheus__Client__Metric **)a);
+  Io__Prometheus__Client__Metric const *m_b =
+      *((Io__Prometheus__Client__Metric **)b);
+
+  if (m_a->n_label < m_b->n_label)
+    return -1;
+  else if (m_a->n_label > m_b->n_label)
+    return 1;
+
+  /* Prometheus does not care about the order of labels. All labels in this
+   * plugin are created by METRIC_ADD_LABELS(), though, and therefore always
+   * appear in the same order. We take advantage of this and simplify the check
+   * by making sure all labels are the same in each position.
+   *
+   * We also only need to check the label values, because the label names are
+   * the same for all metrics in a metric family.
+   *
+   * 3 labels:
+   * [0] $plugin="$plugin_instance" => $plugin is the same within a family
+   * [1] type="$type_instance"      => "type" is a static string
+   * [2] instance="$host"           => "instance" is a static string
```

Comment:
As there already exists several ways to forward metrics from one collectd to another, we can expect some users will setup a central instance collecting values from a fleet of agents, and have prometheus point to this single central collectd.

I agree with you it's better to have collectd hit the "data source" directly, but I'm also aware some people have legacy machines or exotic architectures on which they won't/can't install the latest collectd, nor any prometheus exporter. These will be the typical users of forwarding I guess.